### PR TITLE
restore loc repository before building

### PR DESCRIPTION
**Pull request by OsmondJiang**
- support 4 types of loc mappings, only "Repository" and "RepositoryAndFolder" need restore loc repo before building
- find loc repo and edit url share the same logic
- only restore root docset loc repo

### CODE REVIEW

File: src/docfx/restore/RestoreGit.cs [hunk 1]
```diff
@@ -42,162 +45,66 @@ public static async Task<IEnumerable<(string href, string workTreeHead)>> Restor
...
-                    {
-                        leftWorkTrees.Add(workTree);
-                    }
+                    return default;
                 }
-                await GitUtility.PruneWorkTrees(restorePath);
 
-                return leftWorkTrees;
+                var locRepoFullName = LocConfigConvention.GetLocRepository(config.LocMappingType, repo.FullName, locale, config.DefaultLocale);
+                var locRepo = repo.With(repo.Owner, locRepoFullName.Split(new[] { '\\', '/' }, StringSplitOptions.RemoveEmptyEntries).Last());
```
Comment (live1206) [inline src/docfx/restore/RestoreGit.cs]: >repo.Owner [](start = 40, length = 10)

Updating with the same owner? Seems only need to set Name here, do we need a With method?

File: src/docfx/loc/LocConfigurationConvention.cs [hunk 2]
```diff
@@ -10,17 +10,24 @@ internal static class LocConfigConvention
     {
         private static readonly Regex s_repoNameWithLocale = new Regex(@"^.+?(\.[a-z]{2,4}-[a-z]{2,4}(-[a-z]{2,4})?)?$", RegexOptions.IgnoreCase);
 
+        public static string GetEditRepository(this Config config, string locale)
```
Comment (live1206) [inline src/docfx/loc/LocConfigurationConvention.cs]: >GetEditRepository [](start = 29, length = 17)

What is the difference between GetEditRepository and GetLocRepository? If we put repository to input parameters, we can just expose this extension method?

File: docs/specs/config.yml [hunk 3]
```diff
@@ -508,11 +508,11 @@ commands:
 inputs:
```
Comment (yufeih) [inline docs/specs/config.yml]: how about a dedicated `localization.yml` spec ?

File: docs/specs/config.yml [hunk 4]
```diff
@@ -508,11 +508,11 @@ commands:
...
   docfx.yml: |
     contribution:
-      repository: github.com/docfx/docfx
+      repository: dotnet/docfx
   docs/a.md:
 outputs:
   docs/a.json: |
-    { "content_git_url": "*github.com/docfx/docfx.zh-cn*"}
+    { "content_git_url": "*dotnet/docfx.zh-cn*"}
```
Comment (yufeih) [inline docs/specs/config.yml]: can we remove the wildcard to make the test result more explicit? we only use wildcard only if there are no other choice.

File: src/docfx/loc/LocMappingType.cs [hunk 5]
```diff
@@ -0,0 +1,13 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Docs.Build
+{
+    public enum LocMappingType
+    {
+        Repository, // loc files are stored in an independent repository like dotnet/docfx.zh-cn per locale
```
Comment (yufeih) [inline src/docfx/loc/LocMappingType.cs]: use `/// <summary>` comments. 

The examples in the comments does not state clearly which part is _repo_, which part is _folder_.

For the **Folder** convention, is it
```
/zh-cn/articles/a.md
/zh-tw/articles/a.md
/articles/a.md
```
or
```
/localization/zh-cn/articles/a.md
/localization/zh-tw/articles/a.md
articles/a.md
```
The latter feels better when locale list is large.

Suppose we have **Folder** convention, do we still need **SideBySide** convention? We can just support **Folder** for initial release

File: docs/specs/config.yml (hunk 3 above)
Comment (OsmondJiang) [reply docs/specs/config.yml]: sounds good, changed

---
In reply to: [223994665](https://github.com/dotnet/docfx/pull/3482#discussion_r223994665) [](ancestors = 223994665)

File: docs/specs/config.yml (hunk 4 above)
Comment (OsmondJiang) [reply docs/specs/config.yml]: we do not have other choice here, the content_git_url for spec test contains rooted path which is different for different machine and platform

---
In reply to: [223995143](https://github.com/dotnet/docfx/pull/3482#discussion_r223995143) [](ancestors = 223995143)

File: src/docfx/loc/LocMappingType.cs (hunk 5 above)
Comment (OsmondJiang) [reply src/docfx/loc/LocMappingType.cs]: thanks for the suggestion.

For SideBySide, I believe the ideal is from you, where did you get this idea?

---
In reply to: [223997053](https://github.com/dotnet/docfx/pull/3482#discussion_r223997053) [](ancestors = 223997053)

Comment (yufeih) [reply src/docfx/loc/LocMappingType.cs]: It's from here: https://github.com/dotnet/docfx/issues/803

Comment (OsmondJiang) [reply src/docfx/loc/LocMappingType.cs]: we can just support Folder for initial release, like you said

---
In reply to: [224298010](https://github.com/dotnet/docfx/pull/3482#discussion_r224298010) [](ancestors = 224298010)

File: src/docfx/restore/RestoreGit.cs [hunk 6]
```diff
@@ -42,162 +35,80 @@ public static async Task<IEnumerable<(string href, string workTreeHead)>> Restor
...
             {
-                if (GitUtility.IsRepo(restoreDir))
-                {
-                    // already exists, just pull the new updates from remote
-                    // fetch bare repo: https://stackoverflow.com/questions/3382679/how-do-i-update-my-bare-repo
-                    return GitUtility.Fetch(restorePath, url, "+refs/heads/*:refs/heads/*", token);
-                }
-                else
+                // restore dependency repositories
+                var restoreItems = config.Dependencies.Values.GroupBy(d => GetRestoreRootDir(d), PathUtility.PathComparer).Select(g => (g.Key, g.Distinct().ToList())).ToList();
```
Comment (yufeih) [inline src/docfx/restore/RestoreGit.cs]: can we have a method that returns all git dependencies: `IEnumerable<(string? name, string remote)> GetGitDependencies(Config config, string locale)`

Comment (OsmondJiang) [reply src/docfx/restore/RestoreGit.cs]: already have, RestoreGit.GetRestoreItems()

**Pull request merged**